main/src/main/scala/sbt/Project.scala

```diff
@@ -121,7 +121,44 @@ sealed trait ProjectDefinition[PR <: ProjectReference] {
     if (ts.isEmpty) Nil else s"$label: $ts" :: Nil
 }
 
-sealed trait Project extends ProjectDefinition[ProjectReference] {
+trait CompositeProject {
+  def componentProjects: Seq[Project]
+}
+
+private[sbt] object CompositeProject {
+
+  /**
+   *  Expand user defined `projects` with the component projects of `compositeProjects`.
+   *
+   *  If two projects with the same id appear in the user defined `projects` and
+   *  in `compositeProjects.componentProjects`, the one in `projects` wins.
+   *  This is necessary for backward compatibility with the idioms:
+   *  {{{
+   *    lazy val foo = crossProject
+   *    lazy val fooJS = foo.js.settings(...)
+   *    lazy val fooJVM = foo.jvm.settings(...)
+   *  }}}
+   *  and the rarer:
+   *  {{{
+   *    lazy val fooJS = foo.js.settings(...)
+   *    lazy val foo = crossProject
+   *    lazy val fooJVM = foo.jvm.settings(...)
+   *  }}}
+   */
+  def expand(projects: Seq[Project], compositeProjects: Seq[CompositeProject]): Seq[Project] = {
```

Comment:
You can compute `projects` inside `expand` from `compositeProjects` once and for all, since all projects are also composite projects, so that call sites don't need to know about this:
```scala
val projects = compositeProjects.collect {
  case project: Project => project
}
```
No need to pass it as an additional parameter when `expand` can compute it itself, especially since the call sites are basically `collect` ing as well to give it to `expand` anyway.

Reply:
Yep, so I've removed it - thx :+1: